Single session id for analytics

kgangineni opened the pull request:
Thank you for your contribution to PayPal. 

### Summary of changes
[Jira](https://engineering.paypalcorp.com/jira/browse/DTNOR-644)
 - Single session Id for all FPTI events in same session
 - delegate analytics to AnalyticsClient

 ### Checklist

 - [ ] Added a changelog entry

### Authors
> List GitHub usernames for everyone who contributed to this pull request.
- @kgangineni @scannillo

kgangineni committed:
move analytics code to AnalyticsClient.kt
single session id for all FPTI events

kgangineni committed:
unit tests for AnalyticsClient.kt

kgangineni committed:
Merge branch 'main' into singleSessionIdForAnalytics

kgangineni committed:
remove context from API

kgangineni committed:
remove context from API

sshropshire commented on Core/src/main/java/com/paypal/android/core/analytics/AnalyticsClient.kt:
```diff
@@ -0,0 +1,31 @@
+import com.paypal.android.core.HttpRequestFactory
+import java.util.UUID
+
+internal class AnalyticsClient(
```
q: is the goal to instantiate a new `AnalyticsClient` for each feature client?

scannillo replied:
Similar question [here](https://github.com/paypal/Android-SDK/pull/107#discussion_r1039909042)

kgangineni replied:
@sshropshire moving it to `API`

scannillo commented on Core/src/main/java/com/paypal/android/core/API.kt:
```diff
@@ -1,31 +1,20 @@
 ) {
 
-    constructor(configuration: CoreConfig, context: Context) :
+    constructor(configuration: CoreConfig) :
```
Each feature client will need to still pass along a `Context` at some point.

I think we should keep the `context` as a dependency of `API`, so that each feature client can easily just call `API.sendEvent("card.started")`, for example. That method call, will delegate to an `AnalyticsService` so none of the analytics specific logic lives in `API`. Here's how I thought of doing it on iOS (https://github.com/paypal/iOS-SDK/pull/100).

With this current approach, each feature client would need _both_ an `API` dependency and an `AnalyticsClient` dependency. What do you think?

scannillo committed:
- Re-add context to `API` constructor
- Re-add `API.sendAnalyticsEvent()` method so each feature client does not need an additional `AnalyticsClient`
- Separate out unit tests within AnalyticsClientTest.kt

Co-authored-by: Sammy Cannillo <scannillo@paypal.com>
Co-authored-by: Karthik Gangineni <kgangineni@paypal.com>

scannillo committed:
- Re-add test for API to verify proper call of `AnalyticsClient.sendEvent`

Co-authored-by: Sammy Cannillo <scannillo@paypal.com>
Co-authored-by: Karthik Gangineni <kgangineni@paypal.com>

kgangineni committed:
allow wild card imports in detekt

jcnoriega reviewed (approved):
I wonder if in the future it would be better to have feature client API. For example, something like `CardAPI` where it has `confirmPaymentSource()` and other `Card` related `API` calls. All of this `api` clients could extend `API`, where the `AnalyticsClient` would live.

scannillo commented on Core/src/main/java/com/paypal/android/core/analytics/AnalyticsClient.kt:
```diff
@@ -0,0 +1,30 @@
+import com.paypal.android.core.HttpRequestFactory
+import java.util.UUID
+
+internal class AnalyticsClient(
```
Nitpick / naming comment - I think we use the word `Client` a lot, and it's meaning is a bit conflated. I opted for `AnalyticsService` on iOS, since the android feature modules aren't interfacing directly with this `Analytics` layer & it is what we use on BT.

I don't feel too strongly, but it would be nice if we can align on the same name and decide what makes something a `Client`.

jcnoriega replied:
Agree, perhaps we could reserve `Client` to the classes the the merchant should interact with.

kgangineni committed:
rename AnalyticsClient.kt to AnalyticsService.kt
